Sprint 09

AryunaSan opened the pull request.

AryunaSan committed:
создала экран пользователя в сториборд

AryunaSan committed:
добавила экран SingleImage
добавила Scrollview

AryunaSan committed:
настроила зум на SingleImage

AryunaSan committed:
добавила кнопку Share на экран SingleImage

AryunaSan committed:
добавила рескейл SingleImage

AryunaSan committed:
не работает кнопка share на single image
поправила констрейнты скроллвью
начала верстку кодом страницу пользователя

AryunaSan committed:
добавила кодом три Label, юзерпик имейдж и кнопку логаута в ProfileViewController

AryunaSan committed:
кнопка выхода не отображается на странице профиля

AryunaSan committed:
1. Экран профиля пользователя свёрстан кодом
2, Добавлен экран отдельного изображения SingleImageViewController
3. На экране SingleImageViewController реализована кнопка «Поделиться»:

задание спринт 9 готово к проверке

romanenko-ivan commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,107 @@
+    private var nameLabel: UILabel?
+    private var nicknameLabel: UILabel?
+    private var statusLabel: UILabel?
+    private var userpickImage: UIImageView?
```
Предлагаю более простое и емкое название - `avatarImageView`

romanenko-ivan commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,107 @@
+    // MARK: Private methods
+    
+    private func setUserpickImage() {
+        let userpickImage = UIImage(named: "photo")
```
Давай дадим более предметное название для картинки:

"photo" -> "profileMockImage"

romanenko-ivan commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,107 @@
+    }
+    
+    private func setLogOutButton() {
+        guard let exitImage = UIImage(named: "logOut"),
+              let profileView = self.userpickImage else { return }
+        
+        let logoutButton = UIButton.systemButton(
+            with: exitImage,
+            target: self,
+            action: #selector(Self.didTapButton)
+        )
+        logoutButton.tintColor = .red
+        logoutButton.translatesAutoresizingMaskIntoConstraints = false
+        view.addSubview(logoutButton)
+        
+        logoutButton.widthAnchor.constraint(equalToConstant: 24).isActive = true
+        logoutButton.heightAnchor.constraint(equalToConstant: 24).isActive = true
+        logoutButton.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -24).isActive = true
+        logoutButton.centerYAnchor.constraint(equalTo: profileView.centerYAnchor).isActive = true
+        
+        self.logoutButton = logoutButton
+    }
```
Также, могу предложить кастомизировать свойства прямо при объявлении через замыкание:
```
let topLineView: UIView = {
    let lineView = UIView()
    lineView.layer.borderWidth = 1.0
    lineView.layer.borderColor = UIColor.lightGray.cgColor
    return lineView
}()

```
Таким образом вся настройка компонента будет в одном месте, а в методе, например `configureView()`, выполнить настройку всего лейаута(`addSubview()` и `NSLayoutConstraint.activate([])`)

romanenko-ivan replied:
Аналогично можно сделать для всех элементов на экране "Профиль", а уже все констрейнты установить в одном месте(методе) - например, `configureLayout`

romanenko-ivan commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,107 @@
+        self.logoutButton = logoutButton
+    }
+    @objc
+    private func didTapButton() {}
```
Давай назовем метод более предметно, например, `didTapLogoutButton`

romanenko-ivan commented on ImageFeed/ImagesList/SingleImageViewController.swift:
```diff
@@ -0,0 +1,70 @@
+        dismiss(animated: true, completion: nil)
+    }
+    
+    @IBAction func didTapShareButton(_ sender: UIButton) {
```
Давай добавим самый закрытый модификатор доступа для метода

AryunaSan committed:
на странице профиля пользователя:
1. Исправила цвет фона
2. Исправила цвет текста никнейма
3. Переделала кнопку выхода через обычный UIButton

AryunaSan commented:
@romanenko-ivan Иван, привет! 
отправила пр на повторное ревью. 
спасибо большое за комментарии! 
Долго сидела над комментарием про замыкания. Но только запуталась. поэтому решила пока оставить как есть.
 
У меня вопрос по замечаниям по проекту. Ты пишешь про "Отступ кнопки "Share" от нижнего края должен быть 30px вместо 50px в соотвествии с дизайном", но я не нашла, где именно там 30px. Возможно, вернешь назад проект, тогда буду искать)

А с замыканиями я еще посижу позже)

romanenko-ivan commented:
Про отступ кнопки "Share" - прикладываю скриншот из Фигмы со значением:

<img width="1450" alt="Screenshot 2024-09-27 at 14 52 06" src="https://github.com/user-attachments/assets/959ec66c-a396-4ff2-afd6-8f713130a02a">

romanenko-ivan commented:
Про то, как создавать UI компоненты через замыкания можно прочитать подробнее здесь:
https://medium.com/@rckim77/better-programmatic-components-using-closures-50d1086f8b79

AryunaSan commented:
> Про отступ кнопки "Share" - прикладываю скриншот из Фигмы со значением:
> 
> <img alt="Screenshot 2024-09-27 at 14 52 06" width="1450" src="https://private-user-images.githubusercontent.com/82770629/371531976-959ec66c-a396-4ff2-afd6-8f713130a02a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc0NDA5MDQsIm5iZiI6MTcyNzQ0MDYwNCwicGF0aCI6Ii84Mjc3MDYyOS8zNzE1MzE5NzYtOTU5ZWM2NmMtYTM5Ni00ZmYyLWFmZDYtOGY3MTMxMzBhMDJhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI3VDEyMzY0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZjOTdiNTQyZmMxMjIyNmRlMzY2NTEwODkzNDNhMGUzZWVmN2NjYWNmNGIyZjM4MDM3MTgzNzIxOTljOTNhZjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.9gPwVqZ7_ZYG4yNLOFIbMunjVswcTF0jZStqkvQBmIg">

Иван, я поняла. мы смотрим на разные дизайны. Ты смотришь на 320х568, а я на 375х812. Я правильно понимаю, что мы должны верстать по первому, на который ты даешь ссылку?

romanenko-ivan commented:
> > Про отступ кнопки "Share" - прикладываю скриншот из Фигмы со значением:
> > <img alt="Screenshot 2024-09-27 at 14 52 06" width="1450" src="https://private-user-images.githubusercontent.com/82770629/371531976-959ec66c-a396-4ff2-afd6-8f713130a02a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc0NDA5MDQsIm5iZiI6MTcyNzQ0MDYwNCwicGF0aCI6Ii84Mjc3MDYyOS8zNzE1MzE5NzYtOTU5ZWM2NmMtYTM5Ni00ZmYyLWFmZDYtOGY3MTMxMzBhMDJhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI3VDEyMzY0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZjOTdiNTQyZmMxMjIyNmRlMzY2NTEwODkzNDNhMGUzZWVmN2NjYWNmNGIyZjM4MDM3MTgzNzIxOTljOTNhZjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.9gPwVqZ7_ZYG4yNLOFIbMunjVswcTF0jZStqkvQBmIg">
> 
> Иван, я поняла. мы смотрим на разные дизайны. Ты смотришь на 320х568, а я на 375х812. Я правильно понимаю, что мы должны верстать по первому, на который ты даешь ссылку?

Тогда давай поступим следующим образом, если ты выбрала макет 375х812, то тогда давай следовать его соответствию и выставим отступ в 16px снизу от Safe Area (т.к. по макету отступ в 50px является отступом от самой нижней границы экрана, однако там есть еще "безопасная зона" - Home Indicator, которая тоже занимает место - отсюда и получаем 16px). Прикладываю скрин, чтобы подробнее дать понять о каком констреинте я пишу. 

![image](https://github.com/user-attachments/assets/14471ae4-231b-4d53-88d1-54da58fac35c)

Жду исправлений! :)

AryunaSan committed:
исправила расположение кнопки Share - 16px от Safe Area

AryunaSan commented:
> > > Про отступ кнопки "Share" - прикладываю скриншот из Фигмы со значением:
> > > <img alt="Screenshot 2024-09-27 at 14 52 06" width="1450" src="https://private-user-images.githubusercontent.com/82770629/371531976-959ec66c-a396-4ff2-afd6-8f713130a02a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc0NDA5MDQsIm5iZiI6MTcyNzQ0MDYwNCwicGF0aCI6Ii84Mjc3MDYyOS8zNzE1MzE5NzYtOTU5ZWM2NmMtYTM5Ni00ZmYyLWFmZDYtOGY3MTMxMzBhMDJhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI3VDEyMzY0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZjOTdiNTQyZmMxMjIyNmRlMzY2NTEwODkzNDNhMGUzZWVmN2NjYWNmNGIyZjM4MDM3MTgzNzIxOTljOTNhZjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.9gPwVqZ7_ZYG4yNLOFIbMunjVswcTF0jZStqkvQBmIg">
> > 
> > 
> > Иван, я поняла. мы смотрим на разные дизайны. Ты смотришь на 320х568, а я на 375х812. Я правильно понимаю, что мы должны верстать по первому, на который ты даешь ссылку?
> 
> Тогда давай поступим следующим образом, если ты выбрала макет 375х812, то тогда давай следовать его соответствию и выставим отступ в 16px снизу от Safe Area (т.к. по макету отступ в 50px является отступом от самой нижней границы экрана, однако там есть еще "безопасная зона" - Home Indicator, которая тоже занимает место - отсюда и получаем 16px). Прикладываю скрин, чтобы подробнее дать понять о каком констреинте я пишу.
> 
> ![image](https://private-user-images.githubusercontent.com/82770629/371577729-14471ae4-231b-4d53-88d1-54da58fac35c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc0NTIzOTcsIm5iZiI6MTcyNzQ1MjA5NywicGF0aCI6Ii84Mjc3MDYyOS8zNzE1Nzc3MjktMTQ0NzFhZTQtMjMxYi00ZDUzLTg4ZDEtNTRkYTU4ZmFjMzVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI3VDE1NDgxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFlZGQ3NDJhOTY1NTQzODA4ZWUyOWFlOTUzNWU3YjYwZWE4MTJlODAwMTVkZGRjNjE4ZGYzMjAwYTY3YzQzZTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.CV90c7gqu8LV10jWFph5T0SXaC6u5hfyLvDIvoJAjSE)
> 
> Жду исправлений! :)

Иван, отправила на проверку. Надеюсь, там все правильно. Иначе будет оочень глупо с моей стороны😅
